File: arrow/src/main/scala/org/apache/iceberg/arrow/ArrowUtils.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iceberg.arrow

import org.apache.arrow.memory.RootAllocator
import org.apache.arrow.vector.types.{DateUnit, FloatingPointPrecision, TimeUnit}
import org.apache.arrow.vector.types.pojo.{ArrowType, Field, FieldType, Schema}
import org.apache.spark.sql.internal.SQLConf
import org.apache.spark.sql.types.{ArrayType, BinaryType, BooleanType, ByteType,
  DataType, DateType, DecimalType, DoubleType, FloatType, IntegerType, LongType,
  ShortType, StringType, StructField, StructType, TimestampType}
import scala.collection.JavaConverters._

/**
 * This code is copied from Apache Spark Project's SQL ArrowUtils to be able
  * to shade the Arrow dependencies and keep Arrow code independent from Spark
 */
object ArrowUtils {

  val rootAllocator = new RootAllocator(Long.MaxValue)

  // todo: support more types.
  // scalastyle:off cyclomatic.complexity
  def fromArrowType(dt: ArrowType): DataType = dt match {
    case ArrowType.Bool.INSTANCE => BooleanType
    case int: ArrowType.Int if int.getIsSigned && int.getBitWidth == 8 => ByteType
    case int: ArrowType.Int if int.getIsSigned && int.getBitWidth == 8 * 2 => ShortType
    case int: ArrowType.Int if int.getIsSigned && int.getBitWidth == 8 * 4 => IntegerType
    case int: ArrowType.Int if int.getIsSigned && int.getBitWidth == 8 * 8 => LongType
    case float: ArrowType.FloatingPoint if float.getPrecision() == FloatingPointPrecision.SINGLE => FloatType
    case float: ArrowType.FloatingPoint if float.getPrecision() == FloatingPointPrecision.DOUBLE => DoubleType
    case ArrowType.Utf8.INSTANCE => StringType
    case ArrowType.Binary.INSTANCE => BinaryType
    case d: ArrowType.Decimal => DecimalType(d.getPrecision, d.getScale)
    case date: ArrowType.Date if date.getUnit == DateUnit.DAY => DateType
    case ts: ArrowType.Timestamp if ts.getUnit == TimeUnit.MICROSECOND => TimestampType
    case _ => throw new UnsupportedOperationException(s"Unsupported data type: $dt")
  }
  // scalastyle:on cyclomatic.complexity

  def fromArrowField(field: Field): DataType = {
    field.getType match {
      case ArrowType.List.INSTANCE =>
        val elementField = field.getChildren().get(0)
        val elementType = fromArrowField(elementField)
        ArrayType(elementType, containsNull = elementField.isNullable)
      case ArrowType.Struct.INSTANCE =>
        val fields = field.getChildren().asScala.map { child =>
          val dt = fromArrowField(child)
          StructField(child.getName, dt, child.isNullable)
        }
        StructType(fields)
      case arrowType: ArrowType => fromArrowType(arrowType)
    }
  }

  /** Maps schema from Spark to Arrow. NOTE: timeZoneId required for TimestampType in StructType */
  def fromArrowSchema(schema: Schema): StructType = {
    StructType(schema.getFields.asScala.map { field =>
      val dt = fromArrowField(field)
      StructField(field.getName, dt, field.isNullable)
    })
  }

  /** Return Map with conf settings to be used in ArrowPythonRunner */
  def getPythonRunnerConfMap(conf: SQLConf): Map[String, String] = {
    val timeZoneConf = if (conf.pandasRespectSessionTimeZone) {
      Seq(SQLConf.SESSION_LOCAL_TIMEZONE.key -> conf.sessionLocalTimeZone)
    } else {
      Nil
    }
    val pandasColsByName = Seq(SQLConf.PANDAS_GROUPED_MAP_ASSIGN_COLUMNS_BY_NAME.key ->
      conf.pandasGroupedMapAssignColumnsByName.toString)
    Map(timeZoneConf ++ pandasColsByName: _*)
  }
}

File: spark/src/main/java/org/apache/iceberg/spark/arrow/ArrowUtils.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iceberg.spark.arrow;

import org.apache.arrow.memory.RootAllocator;
import org.apache.arrow.vector.types.DateUnit;
import org.apache.arrow.vector.types.FloatingPointPrecision;
import org.apache.arrow.vector.types.TimeUnit;
import org.apache.arrow.vector.types.pojo.ArrowType;
import org.apache.arrow.vector.types.pojo.Field;
import org.apache.spark.sql.types.ArrayType;
import org.apache.spark.sql.types.DataType;
import org.apache.spark.sql.types.DataTypes;
import org.apache.spark.sql.types.DecimalType;
import org.apache.spark.sql.types.Metadata;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.StructType;

public class ArrowUtils {

  private static ArrowUtils instance;
  private RootAllocator rootAllocator;

  private ArrowUtils() {
    rootAllocator = new RootAllocator(Long.MAX_VALUE);
  }

  public static ArrowUtils instance() {
    if (instance == null) {
      instance = new ArrowUtils();
    }
    return instance;
  }

  public RootAllocator rootAllocator() {
    return rootAllocator;
  }

  @SuppressWarnings("checkstyle:CyclomaticComplexity")
  public DataType fromArrowType(ArrowType data) {

    if (data instanceof ArrowType.Bool) {
      return DataTypes.BooleanType;
    } else if (data instanceof ArrowType.Int) {
      ArrowType.Int intData = (ArrowType.Int) data;
      if (intData.getIsSigned() && intData.getBitWidth() == 8) {
        return DataTypes.ByteType;
      } else if (intData.getIsSigned() && intData.getBitWidth() == 8 * 2) {
        return DataTypes.ShortType;
      } else if (intData.getIsSigned() && intData.getBitWidth() == 8 * 4) {
        return DataTypes.IntegerType;
      } else if (intData.getIsSigned() && intData.getBitWidth() == 8 * 8) {
        return DataTypes.LongType;
      }
    } else if (data instanceof ArrowType.FloatingPoint) {
      ArrowType.FloatingPoint floatData = (ArrowType.FloatingPoint) data;
      if (floatData.getPrecision() == FloatingPointPrecision.SINGLE) {
        return DataTypes.FloatType;
      } else if (floatData.getPrecision() == FloatingPointPrecision.DOUBLE) {
        return DataTypes.DoubleType;
      }
    } else if (data instanceof ArrowType.Utf8) {
      return DataTypes.StringType;
    } else if (data instanceof ArrowType.Binary) {
      return DataTypes.BinaryType;
    } else if (data instanceof ArrowType.Decimal) {
      ArrowType.Decimal decimalData = (ArrowType.Decimal) data;
      return new DecimalType(decimalData.getPrecision(), decimalData.getScale());
    } else if (data instanceof ArrowType.Date && ((ArrowType.Date) data).getUnit() == DateUnit.DAY) {
      return DataTypes.DateType;
    } else if (data instanceof ArrowType.Timestamp && ((ArrowType.Timestamp) data).getUnit() == TimeUnit.MICROSECOND) {
      return DataTypes.TimestampType;
    }

    throw new UnsupportedOperationException("Unsupported data type: " + data);
  }

  public DataType fromArrowField(Field field) {
    ArrowType arrowType = field.getType();
    if (arrowType instanceof ArrowType.List) {
      Field elementField = field.getChildren().get(0);
      DataType elementType = fromArrowField(elementField);
      return new ArrayType(elementType, elementField.isNullable());
    } else if (arrowType instanceof ArrowType.Struct) {
      StructField[] fields = new StructField[field.getChildren().size()];
      int index = 0;
      for (Field f : field.getChildren()) {
        DataType dt = fromArrowField(f);
        fields[index++] = new StructField(f.getName(), dt, f.isNullable(), Metadata.empty());
      }
      return new StructType(fields);
    } else {
      return fromArrowType(arrowType);
    }
  }
}
